### PR TITLE
impl http multi value query str parameters

### PR DESCRIPTION
**Pull request by softprops**
*Issue #, if available:*

*Description of changes:*

this is a follow on to https://github.com/awslabs/aws-lambda-rust-runtime/pull/58 ( which adds support for (de)serializing multi valued headers. since this is a pull against master it will also contain those changes since this was branched from. That pull should be reviewed first and this second.

There will be one final pull to tie the bow and make the lambda-http crate transparently handle both api gateway _and_ alb events. 

By submitting this pull request

- [X] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [X] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

File: lambda-http/src/request.rs [hunk 1]
```diff
@@ -171,8 +223,16 @@ impl<'a> From<GatewayRequest<'a>> for HttpRequest<Body> {
                 path
             )
         });
-
-        builder.extension(QueryStringParameters(query_string_parameters));
+        // multi valued query string parameters are always a super
+        // set of singly valued query string parameters,
+        // when present, multi-valued query string parameters are preferred
+        builder.extension(QueryStringParameters(
```
Comment (softprops) [inline lambda-http/src/request.rs]: abstracting over both cases was a conscious decision. I've found multi valued query string parameter cases awkward to work as two separate collections. the thought process behind making `StrMap` represent both cases was inspired by `HeaderMap` which abstracts over single and multi valued headers.  The semantics here are documented in the public api for users

**Comment by davidbarsky** [comment]
Thanks! Looks solid to me.

**Comment by davidbarsky** [comment]
(There are merge conflicts. I think I merged out of order. I can fix them tomorrow, or you can. Up to you.)

**Comment by softprops** [comment]
pushed conflict resolution commit

**Pull request merged**